parser/lexer.go

```diff
@@ -144,19 +147,20 @@ func (s *Scanner) AppendWarn(err error) {
 	s.warns = append(s.warns, err)
 }
 
-func (s *Scanner) tryDecodeToUTF8String(sql string) string {
-	if mysql.IsUTF8Charset(s.encoding.Name()) {
-		// Skip utf8 encoding because `ToUTF8` validates the whole SQL.
-		// This can cause failure when the SQL contains BLOB values.
-		// TODO: Convert charset on every token and use 'binary' encoding to decode token.
-		return sql
+func (s *Scanner) convert2Connection(tok int, sql string) (int, string) {
+	if mysql.IsUTF8Charset(s.client.Name()) {
+		return tok, sql
 	}
-	utf8Lit, err := s.encoding.Transform(nil, charset.Slice(sql), charset.OpDecodeReplace)
+	utf8Lit, err := s.client.Transform(nil, charset.Slice(sql), charset.OpDecodeReplace)
 	if err != nil {
 		s.AppendError(err)
+		if s.sqlMode.HasStrictMode() && s.client.Tp() == s.connection.Tp() {
+			return invalid, sql
+		}
 		s.lastErrorAsWarn()
 	}
-	return string(utf8Lit)
+	utf8Lit, _ = s.connection.Transform(nil, utf8Lit, charset.OpReplace)
```

Comment:
Why we need this? Could you add a test case to figure it?
Seems if `s.client.Tp() == s.connection.Tp()` we could skip it?

Reply:
MySQL will convert tokens from `client` to `connection`. (see [doc](https://dev.mysql.com/doc/refman/8.0/en/charset-connection.html#:~:text=The%20server%20converts%20statements%20sent%20by%20the%20client%20from%20character_set_client%20to%20character_set_connection)). if `connection` is not `utf8mb4`, we should replace the invalid character with `?` here. 

if `client` and `connection` is the same, we cannot be compatible with MySQL, see the solution https://github.com/pingcap/tidb/issues/29382